From be108b26af411fd41ee0c9e85aab00cddb56573f Mon Sep 17 00:00:00 2001
From: Piyush Gupta <gpiyush@vmware.com>
Date: Tue, 6 Dec 2022 11:01:46 +0000
Subject: [PATCH 14/14] installer.py: Parse string before passing to int()

int() throws ValueError if any string literal is passed to it that
cannot be converted to base 10. Ex - int('112.96k')

Following change adds a method convertToBytes to convert rpm download
size to bytes before initializing package:size dictionary to avoid the
exception.

Change state to 3 if output contains "Testing transaction" to update
message on installer ui.

Change-Id: Icaeddefc65637e84c75b218335b09ee86e87ed96
---
 photon_installer/commandutils.py | 8 ++++++++
 photon_installer/installer.py    | 8 +++++---
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/photon_installer/commandutils.py b/photon_installer/commandutils.py
index d981d44..e03ad8a 100644
--- a/photon_installer/commandutils.py
+++ b/photon_installer/commandutils.py
@@ -130,3 +130,11 @@ class CommandUtils(object):
             self.hostRpmIsNotUsable = 0
 
         return self.hostRpmIsNotUsable
+
+    def convertToBytes(self, size):
+        if not isinstance(size, str):
+            return int(size)
+        if not size[-1].isalpha():
+            return int(size)
+        conv = {'k': 1024, 'm':1024**2, 'g':1024**3, 't':1024**4}
+        return int(float(size[:-1]) * conv[size.lower()[-1]])
diff --git a/photon_installer/installer.py b/photon_installer/installer.py
index 5bc3383..abe38dd 100755
--- a/photon_installer/installer.py
+++ b/photon_installer/installer.py
@@ -988,10 +988,12 @@ class Installer(object):
                     else:
                         info = output.split()
                         package = '{0}-{1}.{2}'.format(info[0], info[2], info[1])
-                        packages_to_install[package] = int(info[5])
-                        total_size += int(info[5])
+                        rpm_download_size = self.cmd.convertToBytes(info[5])
+                        packages_to_install[package] = rpm_download_size
+                        total_size += rpm_download_size
                 elif state == 2:
-                    if output == 'Downloading:\n':
+                    output_status = ["Downloading", "Testing transaction"]
+                    if output.startswith(tuple(output_status)):
                         self.progress_bar.update_message('Preparing ...')
                         state = 3
                 elif state == 3:
-- 
2.23.3

